File: src/main/java/com/demonwav/mcdev/update/ConfigurePluginUpdatesDialog.java
package com.demonwav.mcdev.update;

import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.updateSettings.impl.UpdateSettings;
import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.util.List;

import javax.swing.JComponent;

public class ConfigurePluginUpdatesDialog extends DialogWrapper {

    private ConfigurePluginUpdatesForm form = new ConfigurePluginUpdatesForm();
    private PluginUpdateStatus.Update update;
    private int initialSelectedChannel;

    @SuppressWarnings("unchecked")
    public ConfigurePluginUpdatesDialog() {
        super(true);

        setTitle("Configure Minecraft Development Plugin Updates");
        form.getUpdateCheckInProgressIcon().suspend();
        form.getUpdateCheckInProgressIcon().setPaintPassiveIcon(false);

        form.getChannelBox().addItem("Stable");
        for (Channels channels : Channels.orderedList()) {
            form.getChannelBox().addItem(channels.getTitle());
        }

        form.getCheckForUpdatesNowButton().addActionListener(e -> {
            saveSettings();
            form.getUpdateCheckInProgressIcon().resume();
            resetUpdateStatus();
            PluginUpdater.getInstance().runUpdateCheck(pluginUpdateStatus -> {
                form.getUpdateCheckInProgressIcon().suspend();

                if (pluginUpdateStatus instanceof PluginUpdateStatus.LatestVersionInstalled) {
                    form.getUpdateStatusLabel().setText("You have the latest version of the plugin (" + PluginUtil.getPluginVersion() + ") installed.");
                } else if (pluginUpdateStatus instanceof PluginUpdateStatus.Update) {
                    update = (PluginUpdateStatus.Update) pluginUpdateStatus;
                    form.getInstallButton().setVisible(true);
                    form.getUpdateStatusLabel().setText("A new version (" + ((PluginUpdateStatus.Update) pluginUpdateStatus).getPluginDescriptor().getVersion() + ") is available");
                } else {
                    // CheckFailed
                    form.getUpdateStatusLabel().setText("Update check failed: " + ((PluginUpdateStatus.CheckFailed) pluginUpdateStatus).getMessage());
                }

                return false;
            });
        });

        form.getInstallButton().setVisible(false);
        form.getInstallButton().addActionListener(action -> {
            if (update != null) {
                close(OK_EXIT_CODE);
                try {
                    PluginUpdater.getInstance().installPluginUpdate(update);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        });

        form.getChannelBox().addActionListener(e -> resetUpdateStatus());

        for (Channels channels : Channels.values()) {
            if (channels.hasChannel()) {
                initialSelectedChannel = channels.getIndex();
                break;
            }
        }

        form.getChannelBox().setSelectedIndex(initialSelectedChannel);
        init();
    }

    @Nullable
    @Override
    protected JComponent createCenterPanel() {
        return form.getPanel();
    }

    private void saveSelectedChannel(int channel) {
        final List<String> hosts = UpdateSettings.getInstance().getStoredPluginHosts();
        for (Channels channels : Channels.values()) {
            hosts.remove(channels.getUrl());
        }

        Channels channels = Channels.getChannel(channel);
        if (channels == null) {
            // This really shouldn't happen
            return;
        }

        hosts.add(channels.getUrl());
    }

    private void saveSettings() {
        saveSelectedChannel(form.getChannelBox().getSelectedIndex());
    }

    private void resetUpdateStatus() {
        form.getUpdateStatusLabel().setText(" ");
        form.getInstallButton().setVisible(false);
    }

    @Override
    protected void doOKAction() {
        saveSettings();
        super.doOKAction();
    }

    @Override
    public void doCancelAction() {
        saveSelectedChannel(initialSelectedChannel);
        super.doCancelAction();
    }
}

File: src/main/kotlin/com/demonwav/mcdev/update/ConfigurePluginUpdatesDialog.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.update

import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.updateSettings.impl.UpdateSettings
import java.io.IOException

class ConfigurePluginUpdatesDialog : DialogWrapper(true) {

    private val form = ConfigurePluginUpdatesForm()
    private var update: PluginUpdateStatus.Update? = null
    private var initialSelectedChannel: Int = 0

    init {
        title = "Configure Minecraft Development Plugin Updates"
        form.updateCheckInProgressIcon.suspend()
        form.updateCheckInProgressIcon.setPaintPassiveIcon(false)

        form.channelBox.addItem("Stable")
        for (channels in Channels.values()) {
            form.channelBox.addItem(channels.title)
        }

        form.checkForUpdatesNowButton.addActionListener {
            saveSettings()
            form.updateCheckInProgressIcon.resume()
            resetUpdateStatus()
            PluginUpdater.runUpdateCheck { pluginUpdateStatus ->
                form.updateCheckInProgressIcon.suspend()

                form.updateStatusLabel.text = when (pluginUpdateStatus) {
                    is PluginUpdateStatus.LatestVersionInstalled ->
                        "You have the latest version of the plugin (${PluginUtil.pluginVersion}) installed."
                    is PluginUpdateStatus.Update -> {
                        update = pluginUpdateStatus
                        form.installButton.isVisible = true
                        "A new version (${pluginUpdateStatus.pluginDescriptor.version}) is available"
                    }
                    else -> // CheckFailed
                        "Update check failed: " + (pluginUpdateStatus as PluginUpdateStatus.CheckFailed).message
                }

                false
            }
        }

        form.installButton.isVisible = false
        form.installButton.addActionListener {
            update?.let { update ->
                close(OK_EXIT_CODE)
                try {
                    PluginUpdater.installPluginUpdate(update)
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }

        form.channelBox.addActionListener { resetUpdateStatus() }

        Channels.values().forEachIndexed { i, channel ->
            if (channel.hasChannel()) {
                initialSelectedChannel = i + 1
                return@forEachIndexed
            }
        }

        form.channelBox.selectedIndex = initialSelectedChannel
        init()
    }

    override fun createCenterPanel() = form.panel

    private fun saveSelectedChannel(index: Int) {
        val hosts = UpdateSettings.getInstance().storedPluginHosts
        for (channel in Channels.values()) {
            hosts.remove(channel.url)
        }

        if (index != 0) {
            val channel = Channels.values()[index - 1]
            hosts.add(channel.url)
        }
    }

    private fun saveSettings() {
        saveSelectedChannel(form.channelBox.selectedIndex)
    }

    private fun resetUpdateStatus() {
        form.updateStatusLabel.text = " "
        form.installButton.isVisible = false
    }

    override fun doOKAction() {
        saveSettings()
        super.doOKAction()
    }

    override fun doCancelAction() {
        saveSelectedChannel(initialSelectedChannel)
        super.doCancelAction()
    }
}
